core/src/main/java/org/elasticsearch/index/percolator/PercolatorQueryCache.java

```diff
@@ -135,29 +143,38 @@ QueriesLeaf loadQueries(LeafReaderContext context, Version indexVersionCreated)
         int numQueries = 0;
         Query[] queries = new Query[leafReader.maxDoc()];
 
+        boolean legacyLoading = indexVersionCreated.before(Version.V_5_0_0);
         PostingsEnum postings = leafReader.postings(new Term(TypeFieldMapper.NAME, PercolatorFieldMapper.TYPE_NAME), PostingsEnum.NONE);
         if (postings != null) {
-            for (int docId = postings.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = postings.nextDoc()) {
-                Query query;
-                if (indexVersionCreated.onOrAfter(Version.V_5_0_0)) {
-                    QueryFieldVisitor visitor = new QueryFieldVisitor();
-                    leafReader.document(docId, visitor);
-                    query = parseQueryBuilder(docId, visitor.queryBuilderSource);
-                } else {
-                    LegacyQueryFieldVisitor visitor = new LegacyQueryFieldVisitor();
+            if (legacyLoading) {
+                LegacyQueryFieldVisitor visitor = new LegacyQueryFieldVisitor();
+                for (int docId = postings.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = postings.nextDoc()) {
                     leafReader.document(docId, visitor);
-                    query = parseLegacyPercolatorDocument(docId, visitor.source);
+                    queries[docId] = parseLegacyPercolatorDocument(docId, visitor.source);
+                    numQueries++;
+                    visitor.source = null; // reset
+                }
+            } else {
+                BinaryDocValues binaryDocValues = leafReader.getBinaryDocValues(PercolatorFieldMapper.QUERY_BUILDER_FULL_FIELD_NAME);
+                if (binaryDocValues != null) {
+                    Bits bits = leafReader.getDocsWithField(PercolatorFieldMapper.QUERY_BUILDER_FULL_FIELD_NAME);
+                    for (int docId = postings.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = postings.nextDoc()) {
+                        if (bits.get(docId)) {
+                            BytesRef queryBuilder = binaryDocValues.get(docId);
```

Comment:
the "efficient" way to consume binary doc values would be to do

```java
BytesRef queryBuilder = binaryDocValues.get(docId);
if (queryBuilder.length > 0 || bits.get(docId)) {
    queries[docId] = parseQueryBuilder(docId, queryBuilder);
    numQueries++;
}
```

This works because binary doc values guarantee that docs without a value would return an empty BytesRef.